drivers/spi/spi_renesas_ra.c

```diff
@@ -0,0 +1,826 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_ra_spi
+
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+#include <soc.h>
+#include <instances/r_dtc.h>
+#include <instances/r_spi.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ra_spi);
+
+#include "spi_context.h"
+
+#if defined(CONFIG_SPI_INTERRUPT)
+void spi_rxi_isr(void);
+void spi_txi_isr(void);
+void spi_tei_isr(void);
+void spi_eri_isr(void);
+#endif
+
+struct ra_spi_config {
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct ra_spi_data {
+	struct spi_context ctx;
+	uint8_t dfs;
+	struct st_spi_instance_ctrl spi;
+	struct st_spi_cfg fsp_config;
+	struct st_spi_extended_cfg fsp_config_extend;
+#if CONFIG_SPI_INTERRUPT
+	uint32_t data_len;
+#endif
+#if defined(CONFIG_SPI_RA_DTC)
+	/* RX */
+	struct st_transfer_instance rx_transfer;
+	struct st_dtc_instance_ctrl rx_transfer_ctrl;
+	struct st_transfer_info rx_transfer_info;
+	struct st_transfer_cfg rx_transfer_cfg;
+	struct st_dtc_extended_cfg rx_transfer_cfg_extend;
+
+	/* TX */
+	struct st_transfer_instance tx_transfer;
+	struct st_dtc_instance_ctrl tx_transfer_ctrl;
+	struct st_transfer_info tx_transfer_info;
+	struct st_transfer_cfg tx_transfer_cfg;
+	struct st_dtc_extended_cfg tx_transfer_cfg_extend;
+#endif
+};
+
+static void spi_cb(spi_callback_args_t *p_args)
+{
+	struct device *dev = (struct device *)p_args->p_context;
+	struct ra_spi_data *data = dev->data;
+
+	switch (p_args->event) {
+	case SPI_EVENT_TRANSFER_COMPLETE:
+		spi_context_cs_control(&data->ctx, false);
+		spi_context_complete(&data->ctx, dev, 0);
+		break;
+	case SPI_EVENT_ERR_MODE_FAULT:    /* Mode fault error */
+	case SPI_EVENT_ERR_READ_OVERFLOW: /* Read overflow error */
+	case SPI_EVENT_ERR_PARITY:        /* Parity error */
+	case SPI_EVENT_ERR_OVERRUN:       /* Overrun error */
+	case SPI_EVENT_ERR_FRAMING:       /* Framing error */
+	case SPI_EVENT_ERR_MODE_UNDERRUN: /* Underrun error */
+		spi_context_cs_control(&data->ctx, false);
+		spi_context_complete(&data->ctx, dev, 0);
```

Comment:
this path seems to be the one for errors, but you still return 0 as code? (it should be -EIO)

Reply:
@tbursztyka: Thank you for your feedback, we updated it. Could you please re-check again?